Sources/NetworkProtection/Models/VPNServerSelectionResolver.swift

```diff
@@ -0,0 +1,106 @@
+        self.locationListRepository = locationListRepository
+        self.vpnSettings = vpnSettings
+    }
+
+    public func resolvedServerSelectionMethod() async -> NetworkProtectionServerSelectionMethod {
+        switch currentServerSelectionMethod {
+        case .automatic, .preferredServer, .avoidServer, .failureRecovery:
+            return currentServerSelectionMethod
+        case .preferredLocation(let networkProtectionSelectedLocation):
+            do {
+                let location = try await resolveSelectionAgainstAvailableLocations(networkProtectionSelectedLocation)
+                return .preferredLocation(location)
+            } catch let error as VPNServerSelectionResolverError {
+                switch error {
+                case .countryNotFound:
+                    return .automatic
+                case .fetchingLocationsFailed:
+                    return currentServerSelectionMethod
+                }
+            } catch {
+                return currentServerSelectionMethod
+            }
+        }
+    }
+
+    private func resolveSelectionAgainstAvailableLocations(_ selection: NetworkProtectionSelectedLocation) async throws -> NetworkProtectionSelectedLocation {
+        let availableLocations: [NetworkProtectionLocation]
+        do {
+            availableLocations = try await locationListRepository.fetchLocationListIgnoringCache()
+        } catch {
+            throw VPNServerSelectionResolverError.fetchingLocationsFailed(error)
+        }
+
+        let availableCitySelections = availableLocations.flatMap { location in
+            location.cities.map { city in  NetworkProtectionSelectedLocation(country: location.country, city: city.name) }
+        }
+
+        if availableCitySelections.contains(selection) {
+            return selection
+        }
+
+        let selectedCountry = NetworkProtectionSelectedLocation(country: selection.country)
+        let availableCountrySelections = availableLocations.map { NetworkProtectionSelectedLocation(country: $0.country) }
+        guard availableCountrySelections.contains(selectedCountry) else {
+            throw VPNServerSelectionResolverError.countryNotFound
+        }
+
+        return selectedCountry
+    }
+
+    private var currentServerSelectionMethod: NetworkProtectionServerSelectionMethod {
+        var serverSelectionMethod: NetworkProtectionServerSelectionMethod
+
+        switch vpnSettings.selectedLocation {
+        case .nearest:
+            serverSelectionMethod = .automatic
+        case .location(let networkProtectionSelectedLocation):
+            serverSelectionMethod = .preferredLocation(networkProtectionSelectedLocation)
+        }
+
+        switch vpnSettings.selectedServer {
+        case .automatic:
+            break
+        case .endpoint(let string):
+            // Selecting a specific server will override locations setting
+            // Only available in debug
```

Comment:
Suggestion: Enforce this logic in the code (e.g. guarding `.setSelectedServer(.endpoint(String))` inside `ALPHA || DEBUG`)